### PR TITLE
#50102 Task 2: Add UntilWithSync

### PR DESCRIPTION
**Pull request by tnozicka**
**What this PR does / why we need it**:
This is a split off from https://github.com/kubernetes/kubernetes/pull/50102 to go in smaller pieces.

Introduces UntilWithSync based on informer.

**Needs https://github.com/kubernetes/kubernetes/pull/66906 first**
/hold

**Release note**:
```release-note
NONE
```

/priority important-soon
/kind bug
(bug after the main PR which is this split from)

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/watch/until.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+// UntilWithSync creates an informer from lw, optionally checks precondition when the store is synced,
+// and watches the output until each provided condition succeeds, in a way that is identical
+// to function UntilWithoutRetry. (See above.)
+// UntilWithSync can deal with all errors like API timeout, lost connections and 'Resource version too old'.
+// It is the only function that can recover from 'Resource version too old', Until and UntilWithoutRetry will
+// just fail in that case. On the other hand it can't provide you with guarantees as strong as using simple
+// Watch method with Until. It can skip some intermediate events in case of watch function failing but it will
+// re-list to recover and you always get an event, if there has been a change, after recovery.
+// Also with the current implementation based on DeltaFIFO, order of the events you receive is guaranteed only for
```
Comment (tnozicka) [inline staging/src/k8s.io/client-go/tools/watch/until.go]: @smarterclayton as we were talking about it; the order of events is preserved only for objects with the same key - not really sure why deltaFIFO is implemented that way. Maybe you have some context why it was done like that, if that was intentional or just that for controllers it doesn't matter.

**Comment by tnozicka** [comment]
https://github.com/kubernetes/kubernetes/pull/66906 is in; rebased
/hold cancel

@smarterclayton ptal

**Comment by smarterclayton** [comment]
/approve

but haven't had a chance to go through the PR in detail yet (my previous comments were addressed to the degree I've checked)

**Comment by wojtek-t** [comment]
Thanks for working on that.

/lgtm

**Pull request merged**